plugins/glitch/main.py

```diff
@@ -0,0 +1,53 @@
+import asyncio
+import functools
+import yaml
+
+from .selectors import SELECTOR_MAP as selector_map
+from .actions import ACTION_MAP as action_map
+
+DEFAULT_OUTPUT = '/tmp/last_glitch_plan.yaml'
+
+
+async def glitch(context, rule, action):
+    '''
+
+
+    '''
+    rule.log.info("Starting glitch")
+
+    glitch_output = []
+    output_filename = DEFAULT_OUTPUT
+
+    glitch_plan = action.args.get('glitch_plan') or generate_plan(
+        context.model,
+        num=action.args.get('glitch_num', 5),
+        action_map=action_map
+    )
+    rule.log.info("Writing glitch plan to {}".format(output_file))
+    with open(output_filename, 'w') as output_file:
+        output_file.write(yaml.dump(glitch_plan))
+
+    # Execute glitch plan. We perform destructive operations here!
+    for action in glitch_plan:
+        selectors = action.pop('selectors')
+        actionf = action_map[action.pop('action')]
+
+        # Find a set of units to act upon
+        units = None
+        for selector in selectors:
+            selectf = selector_map[selector.pop['selector']]
+            units = selectf(context.model, **selector)
```

Comment:
I suspect that we can adapt the decorators to deal with things a bit differently here. Though I am open to if/how we do this.

What if for example actions (in particular) were written to operate on a single instance but the decorator included the state machine for things like inflight, interval, max_retries and completor.

@action(inflight=1, interval=5*60)
async def reboot(context, obj):
    await obj.reboot()

which returned the task doing all the interval, retry things.

Likewise,

@action(interval=60, max_retries=3)
async def health(context, obj):
        return all(await obj.healthcheck(), 
                       await obj.agent_state == "idle",
                       await obj.workload_state == "active")

 
Might get called on many units (the whole set at once), each one waiting for a success return, if there are failures it can be retried each interval up to max retries.

The intention here is avoid each action having to include all of this logic (as we already expect them all to operate on sets)

Another way this could work is that @action creates and returns a callable Action instance which manages this more directly. This way we can treat this as a plug point by having @action resolve a factory class (to make it plugable) and by detecting if the action in the plan already points to an Action class. 

What do you think?